### PR TITLE
Allow admin configured mounts to use user configured global credentials

### LINKED ISSUE #2861: External storage: admin should be able to select the global credentials of a user for authentication purposes
## Expected behavior

As an administrator you can choose between different authentication mechanisms if you want to set-up default external storages for your users without even knowing the users password.Therefore it should also be possible to choose `Global credentials of a user` for authentication purposes.

## Current behavior

As an administrator you can choose between the following authentication mechanisms if you set-up default external storages for your users. Unfortunately one option seems to be missing, the `Global credentials of a user` setting so that a user needs to enter the credentials only once and not per share.

2016-12-26 20_36_43-admin - nextcloud
```
...
User entered, store in database -> credentials need to be entered bythe user per share
Global credentials              -> global credentials (entered by the admin)
Global credentials of user      -> missing option!
``` 
### Steps to reproduce
1. Login as an admin
2. Choose Admin -> External storages and check-out the available authentication mechanisms

#### Server Configuration

OS: Linux 3.2.82
Web server: Apache2 2.4.25
Database: MariaDB 5.5.53
PHP version: 5.6.29
Nextcloud version: 10.0.2

#### Client Configuration

Browser: Mozilla Firefox 50.1.0
Operating system: Windows 7

### PR DESCRIPTION
**Pull request by icewind1991**
This is useful when there are multiple admin configured mounts that should all use the same user configured credentials

Fixes #2861

### CODE REVIEW

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+		$this
+			->setIdentifier('password::global::user')
+			->setVisibility(BackendService::VISIBILITY_ADMIN)
+			->setScheme(self::SCHEME_PASSWORD)
+			->setText($l->t('User configured Global credentials'));
+	}
+
+	public function manipulateStorageConfig(StorageConfig &$storage, IUser $user = null) {
+		if (is_null($user)) {
```
Comment (kesselb) [inline apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]:
```suggestion
		if ($user === null) {
```

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			throw new InsufficientDataForMeaningfulAnswerException('No credentials saved');
+		} else {
+			$uid = $user->getUID();
```
Comment (kesselb) [inline apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]: Less complexity without the else

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+use OCA\Files_External\Lib\StorageConfig;
+use OCP\Security\ICredentialsManager;
+use OCA\Files_External\Lib\InsufficientDataForMeaningfulAnswerException;
+
+/**
+ * User provided Global Username and Password
+ */
+class UserGlobalAuth extends AuthMechanism {
+
+	const CREDENTIALS_IDENTIFIER = 'password::global';
```
Comment (kesselb) [inline apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]:
```suggestion
	private const CREDENTIALS_IDENTIFIER = 'password::global';
```

**Review by rullzer — APPROVED** [review]
looks good to me.

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+	protected $credentialsManager;
+
+	public function __construct(IL10N $l, ICredentialsManager $credentialsManager) {
+		$this->credentialsManager = $credentialsManager;
+
+		$this
+			->setIdentifier('password::global::user')
+			->setVisibility(BackendService::VISIBILITY_ADMIN)
+			->setScheme(self::SCHEME_PASSWORD)
+			->setText($l->t('User configured Global credentials'));
```
Comment (wiswedel) [inline apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]: I'd call it "Global credentials, user entered":
- reads better under _Global credentials_
- consistent wording with _User entered, store in database_

image

image

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+	/** @var ICredentialsManager */
+	protected $credentialsManager;
+
+	public function __construct(IL10N $l, ICredentialsManager $credentialsManager) {
+		$this->credentialsManager = $credentialsManager;
+
+		$this
+			->setIdentifier('password::global::user')
+			->setVisibility(BackendService::VISIBILITY_ADMIN)
```
Comment (wiswedel) [inline apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]: ```->setVisibility(BackendService::VISIBILITY_DEFAULT)```

to show it also on the user's list of configured external storages

**Review by wiswedel — CHANGES_REQUESTED** [review]
When credentials haven't been entered or authentication wasn't successful (e.g. after pw change), click on a mount point currently says:

image

Selecting _yes_ leads to

image

Better would be as it is for _User entered, store in database_:

image

**Comment by wiswedel** [comment]
@icewind1991 any feedback to my change requests? Just wanna make sure this goes into 17.

**Comment by rullzer** [comment]
@wiswedel he was on on holiday ;)

**Comment by wiswedel** [comment]
> he was on on holiday ;)

Unbelievable

**Comment by rullzer** [comment]
Master is no 18 development. If this needs to go into 17 please follow the normal backport procedures after merging.

**Comment by wiswedel** [comment]
> Master is no 18 development. If this needs to go into 17 please follow the normal backport procedures after merging.

Ah, an auto-reference by a comment in the 17 beta PR made me assume this was actually supposed to be part of that PR.

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php (hunk 4 above)
Comment (icewind1991) [reply apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]: done

File: apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php (hunk 5 above)
Comment (icewind1991) [reply apps/files_external/lib/Lib/Auth/Password/UserGlobalAuth.php]: done

**Comment by icewind1991** [comment]
You can now configure the credentials as with "user provided"

**Review by ChristophWurst — APPROVED** [review]
Tested & works. Code looks good!

**Comment by wiswedel** [comment]
Thanks a lot, @icewind1991 @rullzer @ChristophWurst @kesselb 
Great work

**Pull request merged**